cmd/registry/provider.go

```diff
@@ -0,0 +1,140 @@
+		Run: func(cmd *cobra.Command, args []string) {
+			var providerName string
+			if len(args) > 0 {
+				providerName = args[0]
+			} else {
+				providerName = input.Text("Enter the name of the provider package to download: ")
+			}
+			addProviderByName(providerName, out)
+		},
+	}
+	cmd.Flags().StringVar(&providerVersion, "version", "latest", "Provider Version to add. Optional")
+	return cmd
+}
+
+func getProviderSearchRoute(providerName string) string {
+	getProviderURL := url.URL{
+		Scheme:   "https",
+		Host:     registryHost,
+		Path:     fmt.Sprintf("%s/%s", registryAPI, providerSearchRoute),
+		RawQuery: fmt.Sprintf(providerSearchQuery, url.QueryEscape(providerName)),
+	}
+	return getProviderURL.String()
+}
+
+func getProviderRoute(providerID, providerVersion string) string {
+	getProviderURL := url.URL{
+		Scheme: "https",
+		Host:   registryHost,
+		Path:   fmt.Sprintf("%s/%s", registryAPI, fmt.Sprintf(providerRoute, providerID, providerVersion)),
+	}
+	return getProviderURL.String()
+}
+
+func addProviderByName(providerName string, out io.Writer) {
+	filledProviderRoute := getProviderSearchRoute(providerName)
+
+	providersJSON := httputil.RequestAndGetJSONBody(filledProviderRoute)
+	providers, exists := providersJSON["providers"].([]interface{})
+	if !exists {
+		jsonString, _ := json.Marshal(providersJSON)
+		log.Fatalf("Couldn't find key 'providers' in Response! %s", jsonString)
+	}
+
+	for _, provider := range providers {
+		childProvidersJSON := provider.(map[string]interface{})
+
+		providerID, childExists := childProvidersJSON["name"].(string) // displayName??
+		if !childExists {
+			jsonString, _ := json.Marshal(childProvidersJSON)
+			log.Fatalf("Couldn't find key 'name' in Response! %s", jsonString)
+		}
+		thisProviderVersion, childExists := childProvidersJSON["version"].(string) // displayName??
+		if !childExists {
+			jsonString, _ := json.Marshal(childProvidersJSON)
+			log.Fatalf("Couldn't find key 'version' in Response! %s", jsonString)
+		}
+		addProviderByIDAndVersion(providerID, thisProviderVersion, out)
+	}
+}
+
+func addProviderByIDAndVersion(providerID, providerVersion string, out io.Writer) {
+	filledProviderRoute := getProviderRoute(providerID, providerVersion)
+	providersJSON := httputil.RequestAndGetJSONBody(filledProviderRoute)
+
+	name, exists := providersJSON["name"].(string)
+	if !exists {
+		jsonString, _ := json.Marshal(providersJSON)
+		log.Fatalf("Couldn't find key 'name' in Response! %s", jsonString)
+	}
+
+	version, exists := providersJSON["version"].(string)
+	if !exists {
+		jsonString, _ := json.Marshal(providersJSON)
+		log.Fatalf("Couldn't find key 'version' in Response! %s", jsonString)
+	}
+	addProviderToRequirementsTxt(name, version, out)
+}
+
+func addProviderToRequirementsTxt(name, version string, out io.Writer) {
+	const filename = "requirements.txt"
+	f, err := os.OpenFile(filename, os.O_APPEND|os.O_WRONLY|os.O_CREATE, writeAndReadPermissions)
+	if err != nil {
+		log.Fatal(err)
+	}
+
```

Comment:
```suggestion
	defer f.Close()
```
nit: file close should be part of defer so that it is called from any return point in the function. Sorry I might have missed it in my earlier review

Reply:
ideally, we shouldn't have log.Fatal(), but rather have it as log.Error() and then return the error from the function, and then the caller should handle the error and act accordingly.

The linter is complaining because as of now f.Close() even if in defer won't be called if there is any log.Fatal() after that in the code execution path. It's not a big issue if we are just missing out on f.Close(), but ideally in Golang we don't exit abruptly and handle errors everywhere, which makes it simpler to maintain in the future.

Reply:
that makes sense. I'll see if I can address that